Use client working directory if supplied for filename completer

puremourning opened the pull request:
Background
====

Fixes: https://github.com/Valloric/YouCompleteMe/issues/1655

The documentation for YouCompleteMe says:

>The g:ycm_filepath_completion_use_working_dir option

> By default, YCM's filepath completion will interpret relative paths like ../ as being relative to the folder of the file of the currently active buffer. Setting this option will force YCM to always interpret relative paths as being relative to Vim's current working directory.

> Default: 0

>  let g:ycm_filepath_completion_use_working_dir = 0

But previously, only the working directory of `ycmd` was used. This meant that operations like changing the working directory of the client were not reflected in completion suggestions.

What this change does
====

The client can now (optionally) pass a `working_dir` property to the '/completions' request which is respected by the filename completer (and potentially any other completer should it be useful). If `working_dir` is not supplied, the previous behaviour remains.

API Changes
====

The `/completions` request data now allows an optional `working_dir` parameter containing the _absolute path_ to the working directory of the client.

Testing
====

Added new tests to the `filename_completer_test`. See also trivial manual test cases on the linked issue. The one remaining question is about the performance of the `os.getcwd` call on each request, but it probably pales into insignificance compared to the performance of the vimscript, and in particular the IO operations performed by the filename completer itself.

This only addresses the 'standard' case for filename completer, as the 'include' case already has its own logic for relative paths and there are already auto tests for that.

Related
====

There is a related (trivial) change to YouCompleteMe: https://github.com/Valloric/YouCompleteMe/pull/1659

vheon commented:
To my understanding we were not following the doc only in the case of a user that has `let g:ycm_filepath_completion_use_working_dir = 1` in his configuration, am I right in this? If that is the case, we could not send the current directory at every request since is not needed.

puremourning commented:
That's certainly true. Arguably it is a slight abuse of the knowledge the client has though. Currently the client doesn't even really know the existence of that switch - it just passes it to the server blindly. OK it's in the docs, so it isn't 100% clear cut, but it is theoretically at least possible for other completers to require (or desire) the working directory of the client. If we hide the data behind the switch, we restrict the server to only having the data in certain configurations which isn't such a clean interface.

Happy to hear others' opinions on this. It's not a difficult change to make.

micbou commented on ycmd/completers/general/tests/filename_completer_test.py:
```diff
@@ -233,3 +236,80 @@ def EnvVar_Undefined_Garbage_3_test( self ):
+        ( 'Qt',       '[Dir]' ),
+        ( 'QtGui',    '[Dir]' ),
+      ], data )
+
```
Missing a blank line.

puremourning replied:
again, sorry I'm not sure where you're expecting another blank line. perhaps there's some style guideline i'm not aware of?!

micbou commented on ycmd/completers/general/tests/filename_completer_test.py:
```diff
@@ -34,6 +34,17 @@
   'file_data' : { PATH_TO_TEST_FILE : { 'filetypes' : [ 'cpp' ] } }
 }
 
```
Missing a blank line.

puremourning replied:
where? I didn't change this line ._0

micbou replied:
Two blank lines are expected before and after functions and classes. You can check this in other files. I think this is to follow the same guideline used in the C++ code. This is also part of the [official style guide](https://www.python.org/dev/peps/pep-0008/#blank-lines) (but we don't respect all of it).

puremourning replied:
Gotya! Thanks.

micbou commented on ycmd/completers/general/tests/filename_completer_test.py:
```diff
@@ -34,6 +34,17 @@
+  candidates = filename_completer.ComputeCandidatesInner( request )
+  return [ ( c[ 'insertion_text' ], c[ 'extra_menu_info' ] )
+          for c in candidates ]
 
```
Missing a blank line.

puremourning replied:
This is copy/pasted and only changed to replace `self` with `filename_completer`

Where are you expecting another line?

micbou commented on ycmd/completers/general/tests/filename_completer_test.py:
```diff
@@ -233,3 +236,80 @@ def EnvVar_Undefined_Garbage_3_test( self ):
+
+    completer = FilenameCompleter( options )
+
+    # change current directory to DATA_DIR/include (path to which we expect 
```
Trailing space.

puremourning replied:
sure will fix

micbou commented:
I would suggest to start comments with a capital letter.

Tests are passing on Windows. I also manually tried it and it was working like expected.

puremourning commented:
Thanks @micbou - I've corrected those errors.

Valloric commented:
> But previously, only the working directory of ycmd was used.

Ha, another relic of when the ycmd code was running inside Vim.

I'm fine with always passing the `working_dir` property. Seems simpler than passing it sometimes.

Agreed with @micbou on capitals for comments.

Valloric commented on ycmd/completers/general/filename_completer.py:
```diff
@@ -133,10 +138,13 @@ def GetPathsIncludeCase( self, path_dir, include_current_file_dir, filepath,
-    path_dir = os.path.join( os.path.dirname( filepath ),
-                             path_dir )
+def _GetPathsStandardCase( path_dir, use_working_dir, filepath, working_dir ):
+  if not path_dir.startswith( '/' ):
```
This IF block smells like a separate function that returns `path_dir`. :)

Valloric commented on ycmd/completers/general/tests/filename_completer_test.py:
```diff
@@ -233,3 +235,81 @@ def EnvVar_Undefined_Garbage_3_test( self ):
     eq_( [ ], data )
+
+
+def WorkingDir_Use_File_Path_test( ):
```
If parens are empty, there should be no space inside them.

Valloric commented:
This is now LGTM from me (conditional on inline comments being resolved, of course).

puremourning commented:
Thanks for the review. I've:

- factored deciding which absolute directory to list into a (local) function. we can expose it should anything else need the logic. Added comments to make it clear how this works.
- corrected the style of `( )` and separation of methods

Valloric commented on ycmd/completers/general/filename_completer.py:
```diff
@@ -133,19 +138,39 @@ def GetPathsIncludeCase( self, path_dir, include_current_file_dir, filepath,
+
+def _GetPathsStandardCase( path_dir, use_working_dir, filepath, working_dir ):
+
+  def GetAbsolutePath( path_dir, use_working_dir, filepath, working_dir ):
```
If it doesn't need to be a inner function, make it a top-level one. The reason is that inner functions have access to all the vars from the enclosing function, and are thus worse at encapsulation. Even if the inner function only uses it's own params, the reader has to audit the whole function to figure that out.

puremourning replied:
OK sure. Done.

micbou commented:
Thanks for fixing all those blank lines. LGTM.

puremourning commented:
OK I think this should be ready now.

Valloric commented:
Thanks!

@homu r+

homu commented:
:pushpin: Commit 9d2d4d0 has been approved by `Valloric`

<!-- @homu r=Valloric 9d2d4d0272ed7ec43bd5e3e067a0e8892df76d35 -->

homu commented:
:hourglass: Testing commit 9d2d4d0 with merge f4a59cf...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/78336332)
